### ISSUE TITLE
Urgent: Custom filter with checkbox and multiselect fields is broken

### ISSUE THREAD

**Issue by dasistwas** [Wunderbyte-GmbH/moodle-mod_datalynx#129]
Problems:

1. Selecting a single value results in the display of all entries
2. Activating checkbox " All selected options have to be part of the entry " automatically adds another tag to the search. Solution: It should not matter if the checkbox is activated or not. Adding an additional tag is a no go even if there is only a single value selected.

**Comment by michaelpollak** [comment]
Can reproduce 1,
with 2 I had no luck. Created a field multiselect (with autocomplete), when I check the "All selected" it works as expected and no tag is added.

Will fix 1 for now, please get back to me with some more info on 2, thank you David.

**Comment by dasistwas** [comment]
2 only concerns field of type multiselect. Zipped video attached.
[multiselect-problem.zip](https://github.com/Wunderbyte-GmbH/moodle-mod_datalynx/files/4681273/multiselect-problem.zip)

**Comment by dasistwas** [comment]
Could you also write a behat test for that? Similar to that one:
https://github.com/Wunderbyte-GmbH/moodle-mod_datalynx/blob/develop/tests/behat/filter_teammember_select.feature.txt
Filter should make sure that following filter using multiselect fields works:

1. Search with a single value selected
2. Search with 2 values selected: (with and without selected checkbox)
3. Combine with fulltext search

**Comment by dasistwas** [comment]
A working behat test is that one. The one mentioned above has outdated syntax:
https://github.com/Wunderbyte-GmbH/moodle-mod_datalynx/blob/develop/tests/behat/fieldgroup_field.feature

**Comment by michaelpollak** [comment]
Got it, can reproduce. Thank you, that's indeed some strange behavior...

**Comment by michaelpollak** [comment]
So, moodle core checks inputs from our forms, it does so with the customfilter and gets the inputs. In this case we send andor plus the selected values (2, 3). In lib/form/select.php:180 we clean the inputs by checking user inputs against options available. 
Options Available: 1, 2, 3, 4
User Input: 2, 3, andor=1

Modform now checks the values against known options and cleans the input. In doing so it just ignores the key "andor" and says puts it on the list, so now we have selected values 1, 2, 3 which is wrong. With "andor = false" we don't run into problems. So after this deep dive into modform and moodle core code I changed the advcheckbox to 0 and a known constant, putting in a text key did not work.

I will look into it some more tomorrow, current testing was successful.

**Comment by dasistwas** [comment]
Search form is autocomplete. Maybe we do not need mod form for search forms?

**Comment by dasistwas** [comment]
Just realised: You were talking about the additional setting for applying AND:
https://docs.moodle.org/dev/lib/formslib.php_Form_Definition#advcheckbox

**Comment by michaelpollak** [comment]
Yes exactly, thank you for looking into that David. The advcheckbox can respond with values other than 0 and 1 which is necessary to work around that weird modform select behavior.

**Comment by dasistwas** [comment]
Still does not work:
Entry with two multiselect values: OTA and OTA OS is not shown when using custom filter and using checkbox  All selected options have to be part of the entry.
Also another bug was introduced.
Also: Please update version.php, this should always be up to date.
Could we cover this in the behat test. I do not see that covered in the the test.

**Comment by dasistwas** [comment]
Clearing form values and then clicking on Search: No entries are shown. Instead all entries should be shown.

**Comment by dasistwas** [comment]
Searching for multiselect value just does not show anything no matter if all selected options have to be part of the entry is chosen.

**Comment by dasistwas** [comment]
Also full text search is broken:
When searching for a multiselect value in full text in custom filter, nothing is shown!

**Comment by dasistwas** [comment]
also teammemberselect search is broken!

**Comment by dasistwas** [comment]
That should not happen. For that we have behat.

**Comment by dasistwas** [comment]
Notice: Undefined offset: 1347 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 1347 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 99 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 99 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 1973 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 1973 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 9571 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 9571 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 10346 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 10346 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 94 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 94 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 4829 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 4829 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 6261 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 6261 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 11311 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'firstname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

Notice: Undefined offset: 11311 in /mod/datalynx/filter/filter_form.php on line 511

Notice: Trying to get property 'lastname' of non-object in /mod/datalynx/filter/filter_form.php on line 511

**Comment by dasistwas** [comment]
I also fixed another issue https://github.com/Wunderbyte-GmbH/moodle-mod_datalynx/commit/ab82b6a4a21143b7ef29e31f6134f10e178489c9 
Please merge.

**Comment by dasistwas** [comment]
Seems the filter is completely broken. Actually I can not do any search using the custom filter. All results are empty

**Comment by michaelpollak** [comment]
The Notices come from users that are authors of an entry but not enrolled in the course at this moment. Do we want to exclude these entries from customfilter or do we want to show their names anyways? Currently only enrolled users are in the authorsearch list, see https://github.com/michaelpollak/moodle-mod_datalynx/blob/71ff585668c285e2a94200f15ab1306133039cea/filter/filter_form.php#L511

**Comment by dasistwas** [comment]
I think the users should be part of the search. Because they could server for administrative use, but without having access to the course.

**Comment by michaelpollak** [comment]
Perfect,
I found the bug, individually everythings works as expected but if I add a search AND I ignore the multiselect it fails. Multiselect probably does not understand when it should be ignored, working on it.

Thanks for the backup David, will get back to you with more infos.

**Comment by dasistwas** [comment]
Did not find this "In Review" strange. Moved back to in progress: I think it is not a good idea to fetch all users. Some Moodle installations have milltions of them!
https://github.com/michaelpollak/moodle-mod_datalynx/blob/8dd30a074578fd0a45347d23d9ba7e9fcaf4d6d3/filter/filter_form.php#L503

**Comment by dasistwas** [comment]
Better would be to extract entry author ids from the datalynx instance. You can find all the ids there: 
https://github.com/michaelpollak/moodle-mod_datalynx/blob/067055a184bbbf0330a16911696bf6cf100c4849/db/install.xml#L167

**Comment by dasistwas** [comment]
Retrieve the ids in an array, then do a list($sql, $params) = $DB->get_in_or_equal (ids)   and then fetch the users from the user table.

**Comment by dasistwas** [comment]
The user ids can be fetched using DISTINCT to get each id only once

**Comment by michaelpollak** [comment]
Thank you, was not yet in review bc. the user list is unpractical and behat is not done. Was just out of time yesterday.
Your solution is way better, thank you for the input, will implement this afternoon after the seminar.

**Comment by dasistwas** [comment]
Finally we got this to work. Thank you for using JOIN makes the solution even more performant.
